File: agents/python-security/agent.py
#!/usr/bin/env python3
"""
Python Security Agent - Cross-implementation proof for A2A protocol

This agent demonstrates that the A2A protocol can be implemented in
any language. It mirrors the TypeScript security agent's functionality.

Run with: uvicorn agent:app --host 127.0.0.1 --port 9210
"""

import re
import uuid
from typing import Any, Literal

from fastapi import FastAPI, Header, HTTPException, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

# =============================================================================
# Configuration
# =============================================================================

PORT = 9210
PROTOCOL_VERSION = "1.0"
SKILL_VERSION = "1.0"

# =============================================================================
# Models
# =============================================================================


class JsonSchemaProperty(BaseModel):
    type: str
    enum: list[str] | None = None
    minimum: int | None = None
    required: list[str] | None = None
    properties: dict[str, Any] | None = None
    items: Any | None = None


class Skill(BaseModel):
    id: str
    version: str
    description: str
    input_schema: JsonSchemaProperty
    output_schema: JsonSchemaProperty


class Auth(BaseModel):
    type: Literal["none", "bearer"]


class AgentCard(BaseModel):
    name: str
    version: str
    protocol_version: str
    endpoint: str
    skills: list[Skill]
    auth: Auth


class Finding(BaseModel):
    severity: Literal["low", "medium", "high", "critical"]
    title: str
    evidence: str
    recommendation: str
    file: str | None = None
    line: int | None = None


class ReviewResult(BaseModel):
    findings: list[Finding]


class InvokeInput(BaseModel):
    diff: str
    mcp_url: str
    additional_context: dict[str, Any] | None = None


class InvokeParams(BaseModel):
    skill: str
    input: InvokeInput


class JsonRpcRequest(BaseModel):
    jsonrpc: Literal["2.0"]
    id: str
    method: str
    params: InvokeParams | None = None


class JsonRpcError(BaseModel):
    code: int
    message: str
    data: Any | None = None


class JsonRpcSuccessResponse(BaseModel):
    jsonrpc: Literal["2.0"] = "2.0"
    id: str
    result: Any


class JsonRpcErrorResponse(BaseModel):
    jsonrpc: Literal["2.0"] = "2.0"
    id: str | None
    error: JsonRpcError


# =============================================================================
# Agent Card
# =============================================================================

AGENT_CARD = AgentCard(
    name="python-security-agent",
    version="0.1",
    protocol_version=PROTOCOL_VERSION,
    endpoint=f"http://127.0.0.1:{PORT}/rpc",
    skills=[
        Skill(
            id="review.security.python",
            version=SKILL_VERSION,
            description="Python-based security analysis for detecting hardcoded secrets",
            input_schema=JsonSchemaProperty(
                type="object",
                required=["diff", "mcp_url"],
                properties={
                    "diff": {"type": "string"},
                    "mcp_url": {"type": "string"},
                },
            ),
            output_schema=JsonSchemaProperty(
                type="object",
                required=["findings"],
                properties={
                    "findings": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "required": ["severity", "title", "evidence", "recommendation"],
                            "properties": {
                                "severity": {"type": "string", "enum": ["low", "medium", "high", "critical"]},
                                "title": {"type": "string"},
                                "evidence": {"type": "string"},
                                "recommendation": {"type": "string"},
                                "file": {"type": "string"},
                                "line": {"type": "integer", "minimum": 1},
                            },
                        },
                    }
                },
            ),
        )
    ],
    auth=Auth(type="none"),
)

# =============================================================================
# Security Analysis
# =============================================================================

# Patterns to detect secrets
SECRET_PATTERNS = [
    (
        re.compile(r'(API_KEY|api_key|apiKey)\s*[=:]\s*["\']([^"\']+)["\']', re.IGNORECASE),
        "API Key",
        "high",
        "Move API keys to environment variables or a secrets manager",
    ),
    (
        re.compile(r'(PASSWORD|password|passwd)\s*[=:]\s*["\']([^"\']+)["\']', re.IGNORECASE),
        "Hardcoded password",
        "critical",
        "Use environment variables or a secrets manager for passwords",
    ),
    (
        re.compile(r'(SECRET|secret|SECRET_KEY|secret_key)\s*[=:]\s*["\']([^"\']+)["\']', re.IGNORECASE),
        "Hardcoded secret",
        "high",
        "Move secrets to environment variables or a secrets manager",
    ),
    (
        re.compile(r'(sk_live_|sk_test_|pk_live_|pk_test_)[a-zA-Z0-9]+'),
        "Stripe API Key",
        "critical",
        "Remove Stripe keys from code; use environment variables",
    ),
    (
        re.compile(r'(ghp_|gho_|ghu_|ghs_|ghr_)[a-zA-Z0-9]+'),
        "GitHub Token",
        "critical",
        "Remove GitHub tokens from code; use environment variables",
    ),
]


def analyze_diff(diff: str) -> list[Finding]:
    """Analyze a diff for security issues."""
    findings: list[Finding] = []
    lines = diff.split("\n")

    current_file: str | None = None
    current_line = 0

    for line in lines:
        # Track file changes
        if line.startswith("+++ b/"):
            current_file = line[6:]
            continue
        elif line.startswith("@@ "):
            # Parse line number from hunk header
            match = re.search(r'\+(\d+)', line)
            if match:
                current_line = int(match.group(1)) - 1
            continue

        # Only check added lines
        if not line.startswith("+") or line.startswith("+++"):
            if line.startswith(" ") or line.startswith("-"):
                current_line += 1
            continue

        current_line += 1
        content = line[1:]  # Remove the + prefix

        # Check each pattern
        for pattern, title, severity, recommendation in SECRET_PATTERNS:
            match = pattern.search(content)
            if match:
                # Extract matched text for evidence
                evidence = f"Found: {match.group(0)}"

                finding = Finding(
                    severity=severity,  # type: ignore
                    title=title,
                    evidence=evidence,
                    recommendation=recommendation,
                    file=current_file,
                    line=current_line,
                )
                findings.append(finding)

    return findings


# =============================================================================
# FastAPI App
# =============================================================================

app = FastAPI(title="Python Security Agent")


@app.get("/.well-known/agent-card.json")
async def get_agent_card() -> AgentCard:
    """Return the Agent Card for discovery."""
    return AGENT_CARD


@app.get("/health")
async def health_check() -> dict[str, str]:
    """Health check endpoint."""
    return {"status": "ok", "agent": "python-security-agent"}


@app.post("/rpc")
async def handle_rpc(
    request: Request,
    authorization: str | None = Header(default=None),
) -> JSONResponse:
    """Handle JSON-RPC 2.0 invoke requests."""
    rpc_id: str | None = None

    try:
        body = await request.json()
    except Exception:
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=None,
                error=JsonRpcError(code=-32700, message="Parse error"),
            ).model_dump(),
        )

    # Validate JSON-RPC envelope
    try:
        rpc_request = JsonRpcRequest(**body)
        rpc_id = rpc_request.id
    except Exception as e:
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=rpc_id,
                error=JsonRpcError(code=-32600, message="Invalid Request", data=str(e)),
            ).model_dump(),
        )

    # Only support "invoke" method
    if rpc_request.method != "invoke":
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=rpc_id,
                error=JsonRpcError(
                    code=-32601,
                    message=f"Method not found: {rpc_request.method}",
                ),
            ).model_dump(),
        )

    # Validate params
    if not rpc_request.params:
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=rpc_id,
                error=JsonRpcError(code=-32602, message="Invalid params: missing params"),
            ).model_dump(),
        )

    # Check skill ID
    if rpc_request.params.skill != "review.security.python":
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=rpc_id,
                error=JsonRpcError(
                    code=-32602,
                    message=f"Unknown skill: {rpc_request.params.skill}. This agent supports: review.security.python",
                ),
            ).model_dump(),
        )

    # Execute the skill
    try:
        findings = analyze_diff(rpc_request.params.input.diff)
        result = ReviewResult(findings=findings)

        return JSONResponse(
            status_code=200,
            content=JsonRpcSuccessResponse(
                id=rpc_id,
                result=result.model_dump(),
            ).model_dump(),
        )
    except Exception as e:
        return JSONResponse(
            status_code=200,
            content=JsonRpcErrorResponse(
                id=rpc_id,
                error=JsonRpcError(code=-32603, message=f"Internal error: {str(e)}"),
            ).model_dump(),
        )


# =============================================================================
# Entry Point
# =============================================================================

if __name__ == "__main__":
    import uvicorn
    print(f"Python Security Agent listening on http://127.0.0.1:{PORT}")
    uvicorn.run(app, host="127.0.0.1", port=PORT, log_level="warning")

File: agents/bun-security/agent.ts
#!/usr/bin/env bun
/**
 * Bun Security Agent - Cross-implementation proof for A2A protocol
 *
 * This agent demonstrates an alternate implementation of the A2A protocol
 * running on Bun, mirroring the TypeScript security agent behavior.
 *
 * Run with: bun run agents/bun-security/agent.ts
 */

type Severity = "low" | "medium" | "high" | "critical";

const PORT = process.env.PORT ? Number.parseInt(process.env.PORT, 10) : 9210;
const PROTOCOL_VERSION = "1.0";
const SKILL_VERSION = "1.0";
const SKILL_ID = "review.security.bun";

const JSON_RPC_ERROR_CODES = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603,
} as const;

const AGENT_CARD = {
  name: "bun-security-agent",
  version: "0.1",
  protocol_version: PROTOCOL_VERSION,
  endpoint: `http://127.0.0.1:${PORT}/rpc`,
  skills: [
    {
      id: SKILL_ID,
      version: SKILL_VERSION,
      description: "Bun-based security analysis for detecting hardcoded secrets",
      input_schema: {
        type: "object",
        required: ["diff", "mcp_url"],
        properties: {
          diff: { type: "string" },
          mcp_url: { type: "string" },
        },
      },
      output_schema: {
        type: "object",
        required: ["findings"],
        properties: {
          findings: {
            type: "array",
            items: {
              type: "object",
              required: ["severity", "title", "evidence", "recommendation"],
              properties: {
                severity: { type: "string", enum: ["low", "medium", "high", "critical"] },
                title: { type: "string" },
                evidence: { type: "string" },
                recommendation: { type: "string" },
                file: { type: "string" },
                line: { type: "integer", minimum: 1 },
              },
            },
          },
        },
      },
    },
  ],
  auth: { type: "none" },
};

interface Finding {
  severity: Severity;
  title: string;
  evidence: string;
  recommendation: string;
  file?: string;
  line?: number;
}

const SECRET_PATTERNS: Array<{
  pattern: RegExp;
  title: string;
  severity: Severity;
  recommendation: string;
}> = [
  {
    pattern: /(API_KEY|api_key|apiKey)\s*[=:]\s*["']([^"']+)["']/i,
    title: "API Key",
    severity: "high",
    recommendation: "Move API keys to environment variables or a secrets manager",
  },
  {
    pattern: /(PASSWORD|password|passwd)\s*[=:]\s*["']([^"']+)["']/i,
    title: "Hardcoded password",
    severity: "critical",
    recommendation: "Use environment variables or a secrets manager for passwords",
  },
  {
    pattern: /(SECRET|secret|SECRET_KEY|secret_key)\s*[=:]\s*["']([^"']+)["']/i,
    title: "Hardcoded secret",
    severity: "high",
    recommendation: "Move secrets to environment variables or a secrets manager",
  },
  {
    pattern: /(sk_live_|sk_test_|pk_live_|pk_test_)[a-zA-Z0-9]+/,
    title: "Stripe API Key",
    severity: "critical",
    recommendation: "Remove Stripe keys from code; use environment variables",
  },
  {
    pattern: /(ghp_|gho_|ghu_|ghs_|ghr_)[a-zA-Z0-9]+/,
    title: "GitHub Token",
    severity: "critical",
    recommendation: "Remove GitHub tokens from code; use environment variables",
  },
];

function jsonResponse(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data), {
    status,
    headers: {
      "Content-Type": "application/json",
    },
  });
}

function jsonRpcError(
  id: string | null,
  code: number,
  message: string,
  data?: unknown,
): Response {
  const error = data === undefined ? { code, message } : { code, message, data };
  return jsonResponse({ jsonrpc: "2.0", id, error });
}

function jsonRpcSuccess(id: string, result: unknown): Response {
  return jsonResponse({ jsonrpc: "2.0", id, result });
}

function analyzeDiff(diff: string): Finding[] {
  const findings: Finding[] = [];
  const lines = diff.split("\n");

  let currentFile: string | null = null;
  let currentLine = 0;

  for (const line of lines) {
    if (line.startsWith("+++ b/")) {
      currentFile = line.slice(6);
      continue;
    }

    const hunkMatch = line.match(/^@@ -\d+(?:,\d+)? \+(\d+)/);
    if (hunkMatch) {
      currentLine = Number.parseInt(hunkMatch[1], 10) - 1;
      continue;
    }

    if (line.startsWith("+") && !line.startsWith("+++")) {
      currentLine += 1;
      const content = line.slice(1);
      for (const { pattern, title, severity, recommendation } of SECRET_PATTERNS) {
        const match = pattern.exec(content);
        if (match) {
          findings.push({
            severity,
            title,
            evidence: `Found: ${match[0]}`,
            recommendation,
            file: currentFile ?? undefined,
            line: currentLine,
          });
        }
      }
      continue;
    }

    if (line.startsWith(" ")) {
      currentLine += 1;
    }
  }

  return findings;
}

const server = Bun.serve({
  port: PORT,
  hostname: "127.0.0.1",
  async fetch(req) {
    const url = new URL(req.url);

    if (req.method === "GET" && url.pathname === "/.well-known/agent-card.json") {
      return jsonResponse(AGENT_CARD);
    }

    if (url.pathname === "/health") {
      return jsonResponse({ status: "ok", agent: "bun-security-agent" });
    }

    if (req.method === "POST" && url.pathname === "/rpc") {
      let body: unknown;
      try {
        body = await req.json();
      } catch {
        return jsonRpcError(null, JSON_RPC_ERROR_CODES.PARSE_ERROR, "Parse error");
      }

      if (
        !body ||
        typeof body !== "object" ||
        (body as { jsonrpc?: string }).jsonrpc !== "2.0" ||
        typeof (body as { id?: unknown }).id !== "string" ||
        typeof (body as { method?: unknown }).method !== "string"
      ) {
        return jsonRpcError(null, JSON_RPC_ERROR_CODES.INVALID_REQUEST, "Invalid Request");
      }

      const rpc = body as {
        id: string;
        method: string;
        params?: {
          skill?: string;
          input?: { diff?: string; mcp_url?: string };
        };
      };

      if (rpc.method !== "invoke") {
        return jsonRpcError(rpc.id, JSON_RPC_ERROR_CODES.METHOD_NOT_FOUND, `Method not found: ${rpc.method}`);
      }

      if (
        !rpc.params ||
        rpc.params.skill !== SKILL_ID ||
        !rpc.params.input ||
        typeof rpc.params.input.diff !== "string" ||
        typeof rpc.params.input.mcp_url !== "string"
      ) {
        return jsonRpcError(rpc.id, JSON_RPC_ERROR_CODES.INVALID_PARAMS, "Invalid params");
      }

      try {
        const findings = analyzeDiff(rpc.params.input.diff);
        return jsonRpcSuccess(rpc.id, { findings });
      } catch (error) {
        const message = error instanceof Error ? error.message : "Internal error";
        return jsonRpcError(rpc.id, JSON_RPC_ERROR_CODES.INTERNAL_ERROR, message);
      }
    }

    return new Response("Not Found", { status: 404 });
  },
});

console.log(`STARTED:${JSON.stringify({ name: "Bun Security Agent", port: server.port })}`);
